README.md
# rn-svg-charts

[![version](https://img.shields.io/npm/v/rn-svg-charts.svg)](https://www.npmjs.com/package/rn-svg-charts)
[![downloads](https://img.shields.io/npm/dm/rn-svg-charts.svg)](https://www.npmjs.com/package/rn-svg-charts)
[![license](https://img.shields.io/npm/l/rn-svg-charts.svg)](https://www.npmjs.com/package/rn-svg-charts)

Welcome to rn-svg-charts!

`rn-svg-charts` provides SVG Charts support to React Native on iOS and Android, and a compatibility layer for the web.

### Looking for maintainers! I alone don't have the time to maintain this library anymore. Preferably looking for somebody who uses this library in their proffesional work (how I originally got the time to maintain).

### version 5 is now available!

A much improved decorator system has been introduced, allowing for greater flexibility and less complexity.
See [releases](https://github.com/mdmush/rn-svg-charts/releases) for more information.

---

In order to not bloat this README too much we've moved some examples over to
[`rn-svg-charts-examples`](https://github.com/mdmush/rn-svg-charts-examples).
There we will try to showcase the really cool things you can do with this library.
This README will try to keep things as simple as possible so that everybody can get up and running as fast as possible.

## Prerequisites

This library uses [react-native-svg](https://github.com/react-native-community/react-native-svg)
to render its graphs. Therefore this library needs to be installed **AND** linked into your project to work.

Other than the above dependency this library uses pure javascript and supports both iOS and Android

## Getting Started

> Make sure to follow the installation mentioned in the [Prerequisites](#Prerequisites)

Then you can install the library using either yarn:

```
yarn add rn-svg-charts
```

or npm:

```
npm install --save rn-svg-charts
```

Now you're good to go!

## Motivation

Creating beautiful graphs in React Native shouldn't be hard or require a ton of knowledge.
We use [react-native-svg](https://github.com/react-native-community/react-native-svg) in order to render our SVG's and to provide you with great extensibility.
We utilize the very popular [d3](https://d3js.org/) library to create our SVG paths and to calculate the coordinates.

We built this library to be as extensible as possible while still providing you with the most common charts and data visualization tools out of the box.
We're very proud of our "decorator" support. All charts can be extended with "decorators", a component that somehow styles or enhances your chart.
Simply pass in a `react-native-svg` compliant component as a child to the graph and it will be called with all the necessary information to layout your decorator.
See each chart for information on what data the decorator will be called with.

Feedback and PR's are more than welcome 🙂

## Running

If you want to test out the library you can clone this repository and run it.
We suggest that you test out the [storybooks](https://github.com/storybooks/storybook) that we've implemented.
Most of the charts are implemented with [knobs](https://github.com/storybooks/storybook/tree/master/addons/knobs) so that you can tweak most properties and see their behavior live.

Clone the repo and run the following:

```bash
yarn

# for iOS
(cd ios && pod install)
react-native run-ios

# for Android
react-native run-android

yarn storybook

# and then reload your device
```

## Common Props

| Property          | Default                                  | Description                                                                                                                                                                                                                                                    |
| ----------------- | ---------------------------------------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| data              | **required**                             | An array of arbitrary data - use prop `xAccessor`/`yAccessor`to tell the chart about the data structure                                                                                                                                                        |
| yAccessor         | ({ item }) => item                       | A function that takes each entry of `data` (named "item") as well as the index and returns the y-value of that entry                                                                                                                                           |
| xAccessor         | ({ index }) => index                     | Same as `yAccessor` but returns the x-value of that entry                                                                                                                                                                                                      |
| yScale            | d3Scale.scaleLinear                      | A function that determines the scale of said axis (only tested with scaleLinear, scaleTime & scaleBand )                                                                                                                                                       |
| xScale            | d3Scale.scaleLinear                      | Same as `yScale` but for the x axis                                                                                                                                                                                                                            |
| svg               | `{}`                                     | an object containing all the props that should be passed down to the underlying `react-native-svg` component. [See available props](https://github.com/react-native-community/react-native-svg#common-props)                                                   |
| animate           | false                                    | PropTypes.bool                                                                                                                                                                                                                                                 |
| animationDuration | 300                                      | PropTypes.number                                                                                                                                                                                                                                               |
| style             | undefined                                | Supports all [ViewStyleProps](https://facebook.github.io/react-native/docs/view-style-props)                                                                                                                                                            |
| curve             | d3.curveLinear                           | A function like [this](https://github.com/d3/d3-shape#curves)                                                                                                                                                                                                  |
| contentInset      | { top: 0, left: 0, right: 0, bottom: 0 } | An object that specifies how much fake "margin" to use inside of the SVG canvas. This is particularly helpful on Android where `overflow: "visible"` isn't supported and might cause clipping. Note: important to have same contentInset on axis's and chart   |
| numberOfTicks     | 10                                       | We use [d3-array](https://github.com/d3/d3-array#ticks) to evenly distribute the grid and dataPoints on the yAxis. This prop specifies how many "ticks" we should try to render. Note: important that this prop is the same on both the chart and on the yAxis |
| showGrid          | true                                     | Whether or not to show the grid lines                                                                                                                                                                                                                          |
| yMin              | undefined                                | Alter how the chart bounds are calculated                                                                                                                                                                                                                      |
| yMax              | undefined                                | Alter how the chart bounds are calculated                                                                                                                                                                                                                      |
| xMin              | undefined                                | Alter how the chart bounds are calculated                                                                                                                                                                                                                      |
| xMax              | undefined                                | Alter how the chart bounds are calculated                                                                                                                                                                                                                      |
| children          | undefined                                | One or many `react-native-svg` components that will be used to enhance your chart                                                                                                                                                                              |

## Common arguments to children

| Property | Description                                                                                                             |
| -------- | ----------------------------------------------------------------------------------------------------------------------- |
| x        | a function that normally accepts the index of a data point an returns its 'x' location on the canvas                    |
| y        | a function that normally accepts the value of a data point an returns its 'y' location on the canvas                    |
| width    | the width of the canvas in pixels                                                                                       |
| height   | the height of the canvas in pixels                                                                                      |
| data     | the same data array provided to the chart, use this to map over your data points if you want decorators on each point   |
| ticks    | if `numberOfTicks` has been provided to the chart this array will include the calculated tick values (useful for grids) |

## Components

This library currently provides the following components

-   [Area](#areachart)
-   [StackedAreaChart](#stackedareachart)
-   [Bar](#barchart)
-   [StackedBarChart](#stackedbarchart)
-   [Line](#linechart)
-   [Pie](#piechart)
-   [ProgressCircle](#progresscircle)
-   [YAxis](#yaxis)
-   [XAxis](#xaxis)

Also see

-   [Children](#children)
-   [Grid](#grid)

### AreaChart

![Area chart](https://raw.githubusercontent.com/mdmush/rn-svg-charts/master/screenshots/area-chart.png)

#### Example

```jsx
import React from 'react'
import { AreaChart, Grid } from 'rn-svg-charts'
import * as shape from 'd3-shape'

class AreaChartExample extends React.PureComponent {
    render() {
        const data = [50, 10, 40, 95, -4, -24, 85, 91, 35, 53, -53, 24, 50, -20, -80]

        return (
            <AreaChart
                style={{ height: 200 }}
                data={data}
                contentInset={{ top: 30, bottom: 30 }}
                curve={shape.curveNatural}
                svg={{ fill: 'rgba(134, 65, 244, 0.8)' }}
            >
                <Grid />
            </AreaChart>
        )
    }
}
```

#### Props

See [Common Props](#common-props)

| Property | Default | Description                                                                  |
| -------- | ------- | ---------------------------------------------------------------------------- |
| start    | 0       | The value of which the area should start (will always end on the data point) |

#### Arguments to children

Supports all [Common arguments to children](#common-arguments-to-children)

### StackedAreaChart

Very similar to an area chart but with multiple sets of data stacked together.
We suggest that you read up on [d3 stacks](https://github.com/d3/d3-shape#stacks) in order to better understand this chart and its props
See [Area stack chart with Y axis](https://github.com/mdmush/rn-svg-charts-examples/blob/master/storybook/stories/area-stack/with-y-axis.js) to see how to use a YAxis with this component
Use the `svgs` prop to pass in `react-native-svg` compliant props to each area.

![Stacked area chart](https://raw.githubusercontent.com/mdmush/rn-svg-charts/master/screenshots/area-stack.png)

#### Example

```jsx
import React from 'react'
import { StackedAreaChart } from 'rn-svg-charts'
import * as shape from 'd3-shape'

class StackedAreaExample extends React.PureComponent {
    render() {
        const data = [
            {
                month: new Date(2015, 0, 1),
                apples: 3840,
                bananas: 1920,
                cherries: 960,
                dates: 400,
            },
            {
                month: new Date(2015, 1, 1),
                apples: 1600,
                bananas: 1440,
                cherries: 960,
                dates: 400,
            },
            {
                month: new Date(2015, 2, 1),
                apples: 640,
                bananas: 960,
                cherries: 3640,
                dates: 400,
            },
            {
                month: new Date(2015, 3, 1),
                apples: 3320,
                bananas: 480,
                cherries: 640,
                dates: 400,
            },
        ]

        const colors = ['#8800cc', '#aa00ff', '#cc66ff', '#eeccff']
        const keys = ['apples', 'bananas', 'cherries', 'dates']
        const svgs = [
            { onPress: () => console.log('apples') },
            { onPress: () => console.log('bananas') },
            { onPress: () => console.log('cherries') },
            { onPress: () => console.log('dates') },
        ]

        return (
            <StackedAreaChart
                style={{ height: 200, paddingVertical: 16 }}
                data={data}
                keys={keys}
                colors={colors}
                curve={shape.curveNatural}
                showGrid={false}
                svgs={svgs}
            />
        )
    }
}
```

#### Props

| Property | Default                                                              | Description                                                               |
| -------- | -------------------------------------------------------------------- | ------------------------------------------------------------------------- |
| data     | **required**                                                         | An array of the data entries                                              |
| keys     | **required**                                                         | This array should contain the object keys of interest (see above example) |
| colors   | **required**                                                         | An array of equal size as `keys` with the color for each key              |
| order    | [d3.stackOrderNone](https://github.com/d3/d3-shape#stackOrderNone)   | The order in which to sort the areas                                      |
| offset   | [d3.stackOffsetNone](https://github.com/d3/d3-shape#stackOffsetNone) | A function to determine the offset of the areas                           |

Also see [Common Props](#common-props)

#### Arguments to children

| Property | Description                                                                                                              |
| -------- | ------------------------------------------------------------------------------------------------------------------------ |
| x        | a function that normally accepts the index of a data points an returns its 'x' location on the canvas                    |
| y        | a function that normally accepts the value of a data points an returns its 'y' location on the canvas                    |
| width    | the width of the canvas in pixels                                                                                        |
| height   | the height of the canvas in pixels                                                                                       |
| ~~data~~ | ~~the same data array provided to the chart, use this to map over your data points if you want decorators on each point~ |
| ticks    | if `numberOfTicks` has been provided to the chart this array will include the calculated tick values (useful for grids)  |

This chart does not call a child with the `data` argument. This is due to the fact that a lot of calculations go into
creating the stacked chart, meaning that the original `data` prop doesn't provide especially valuable information
when trying to layout decorators. It does however call with the rest of the [common arguments](#common-arguments-to-children)

### BarChart

![Bar chart](https://raw.githubusercontent.com/mdmush/rn-svg-charts/master/screenshots/bar-chart.png)

#### Example

```jsx
import React from 'react'
import { BarChart, Grid } from 'rn-svg-charts'

class BarChartExample extends React.PureComponent {
    render() {
        const fill = 'rgb(134, 65, 244)'
        const data = [50, 10, 40, 95, -4, -24, null, 85, undefined, 0, 35, 53, -53, 24, 50, -20, -80]

        return (
            <BarChart style={{ height: 200 }} data={data} svg={{ fill }} contentInset={{ top: 30, bottom: 30 }}>
                <Grid />
            </BarChart>
        )
    }
}
```

### Props

Also see [Common Props](#common-props)

| Property     | Default                                    | Description                                                                                                                                                                                                                                                                                                                                                                             |
| ------------ | ------------------------------------------ | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| data         | **required**                               | The data prop in a barChart can look exactly like in a Line- or AreaChart, i.e an array of just numbers or complex objects. It can however also be an array with several data sets. A data object can contain a `svg` property which allows you two override styles on that specific object. See [the examples repo](https://github.com/mdmush/rn-svg-charts-examples) |
| horizontal   | false                                      | Boolean whether or not the bars should be horizontal                                                                                                                                                                                                                                                                                                                                    |
| svg          | `{}`                                       | Default svg props **for all bars**. Supports all svg props an svg path normally supports. This styles will be overriden if there are specific styles for a given data object                                                                                                                                                                                                            |
| spacingInner | 0.05                                       | Spacing between the bars (or groups of bars)                                                                                                                                                                                                                                                                                                                                            |
| spacingOuter | 0.05                                       | Spacing outside of the bars (or groups of bars). Percentage of one bars width                                                                                                                                                                                                                                                                                                           |
| contentInset | `{ top: 0, left: 0, right: 0, bottom: 0 }` | PropTypes.shape                                                                                                                                                                                                                                                                                                                                                                         |

#### Arguments to children

| Property  | Description                     |
| --------- | ------------------------------- |
| bandwidth | the width of a band (a.k.a bar) |

Also supports all [Common arguments to children](#common-arguments-to-children)

### StackedBarChart

The same as the [StackedAreaChart](#stackedareachart) except with bars (and different `svgs` prop).
We suggest that you read up on [d3 stacks](https://github.com/d3/d3-shape#stacks) in order to better understand this chart and its props

The `svgs` prop here is not based on keys, but rather entries, as the user might want to specify different props
for each entry in each bar. Therefore each key entry can contain a complex object that contains e.g an `svg` prop. See [this example](./storybook/stories/bar-stack/with-on-press.js) for inspiration

![Stacked bar chart](https://raw.githubusercontent.com/mdmush/rn-svg-charts/master/screenshots/bar-stack.png)
![Stacked bar chart - horizontal](https://raw.githubusercontent.com/mdmush/rn-svg-charts/master/screenshots/bar-stack-horizontal.png)

#### Example

```jsx
import React from 'react'
import { StackedBarChart } from 'rn-svg-charts'

class StackedBarChartExample extends React.PureComponent {
    render() {
        const data = [
            {
                month: new Date(2015, 0, 1),
                apples: 3840,
                bananas: 1920,
                cherries: 960,
                dates: 400,
                oranges: 400,
            },
            {
                month: new Date(2015, 1, 1),
                apples: 1600,
                bananas: 1440,
                cherries: 960,
                dates: 400,
            },
            {
                month: new Date(2015, 2, 1),
                apples: 640,
                bananas: 960,
                cherries: 3640,
                dates: 400,
            },
            {
                month: new Date(2015, 3, 1),
                apples: 3320,
                bananas: 480,
                cherries: 640,
                dates: 400,
            },
        ]

        const colors = ['#7b4173', '#a55194', '#ce6dbd', '#de9ed6']
        const keys = ['apples', 'bananas', 'cherries', 'dates']

        return (
            <StackedBarChart
                style={{ height: 200 }}
                keys={keys}
                colors={colors}
                data={data}
                showGrid={false}
                contentInset={{ top: 30, bottom: 30 }}
            />
        )
    }
}
```

#### Props

| Property      | Default                                                              | Description                                                                                                      |
| ------------- | -------------------------------------------------------------------- | ---------------------------------------------------------------------------------------------------------------- |
| data          | **required**                                                         | An array of the data entries: each value can be a number or a complex object with custom `svg` props for example |
| keys          | **required**                                                         | This array should contain the object keys of interest (see above example)                                        |
| colors        | **required**                                                         | An array of equal size as `keys` with the color for each key                                                     |
| valueAccessor | ({ item, key }) => item[key]                                         | Very similar to the `yAccessor` of the other charts, usually needed when using complex objects as values         |
| horizontal    | false                                                                | Boolean whether or not the bars should be horizontal                                                             |
| order         | [d3.stackOrderNone](https://github.com/d3/d3-shape#stackOrderNone)   | The order in which to sort the areas                                                                             |
| offset        | [d3.stackOffsetNone](https://github.com/d3/d3-shape#stackOffsetNone) | A function to determine the offset of the areas                                                                  |

Also see [Common Props](#common-props)

#### Arguments to children

| Property | Description                                                                                                               |
| -------- | ------------------------------------------------------------------------------------------------------------------------- |
| x        | a function that normally accepts the index of a data points an returns its 'x' location on the canvas                     |
| y        | a function that normally accepts the value of a data points an returns its 'y' location on the canvas                     |
| width    | the width of the canvas in pixels                                                                                         |
| height   | the height of the canvas in pixels                                                                                        |
| ~~data~~ | ~~the same data array provided to the chart, use this to map over your data points if you want decorators on each point~~ |
| ticks    | if `numberOfTicks` has been provided to the chart this array will include the calculated tick values (useful for grids)   |

This chart does not call a child with the `data` argument. This is due to the fact that a lot of calculations go into
creating the stacked chart, meaning that the original `data` prop doesn't provide especially valuable information
when trying to layout decorators. It does however call with the rest of the [common arguments](#common-arguments-to-children)

### LineChart

![Line chart](https://raw.githubusercontent.com/mdmush/rn-svg-charts/master/screenshots/line-chart.png)

#### Example

```jsx
import React from 'react'
import { LineChart, Grid } from 'rn-svg-charts'

class LineChartExample extends React.PureComponent {
    render() {
        const data = [50, 10, 40, 95, -4, -24, 85, 91, 35, 53, -53, 24, 50, -20, -80]

        return (
            <LineChart
                style={{ height: 200 }}
                data={data}
                svg={{ stroke: 'rgb(134, 65, 244)' }}
                contentInset={{ top: 20, bottom: 20 }}
            >
                <Grid />
            </LineChart>
        )
    }
}
```

#### Props

See [Common Props](#common-props)

#### Arguments to children

Supports all [Common arguments to children](#common-arguments-to-children)

### PieChart

The PieChart is a really nice component with great support for custom behavior.
See more examples in the [examples repo](https://github.com/mdmush/rn-svg-charts-examples)

![Pie chart](https://raw.githubusercontent.com/mdmush/rn-svg-charts/master/screenshots/pie-chart.png)

#### Example

```jsx
import React from 'react'
import { PieChart } from 'rn-svg-charts'

class PieChartExample extends React.PureComponent {
    render() {
        const data = [50, 10, 40, 95, -4, -24, 85, 91, 35, 53, -53, 24, 50, -20, -80]

        const randomColor = () => ('#' + ((Math.random() * 0xffffff) << 0).toString(16) + '000000').slice(0, 7)

        const pieData = data
            .filter((value) => value > 0)
            .map((value, index) => ({
                value,
                svg: {
                    fill: randomColor(),
                    onPress: () => console.log('press', index),
                },
                key: `pie-${index}`,
            }))

        return <PieChart style={{ height: 200 }} data={pieData} />
    }
}
```

#### Props


CHANGELOG.md
# 5.3.0

### Contributions

-   update README ([@andrewdazs](https://github.com/andrewdazs) and [@iammosespaulr](https://github.com/iammosespaulr))
-   fix crash in Axis component ([@krzysztof-miemiec](https://github.com/krzysztof-miemiec) - [#238](https://github.com/mdmush/rn-svg-charts/pull/238))
-   Introduce grouped LineChart ([@alburdette619](https://github.com/alburdette619) - [#240](https://github.com/mdmush/rn-svg-charts/pull/240))
-   Introduce grouped StackedBarChart ([@alburdette619](https://github.com/alburdette619) - #239)
-   Fix 'transparent' color for X-Axis and Y-Axis in RN 0.57 ([@denieler](https://github.com/denieler) - #256)
-   Added extra props to axis children ([@alburdette619](https://github.com/alburdette619) - #276)
-   Create interaction handle during animation ([@Jyrno42](https://github.com/Jyrno42) - #314)
-   Stacked Area Chart: Pass `areas` down to children ([@attitude](https://github.com/attitude) - #316)
-   Added more props to axis children ([@buschco](https://github.com/buschco) - #326)
-   Fix blinking animation ([@andycloke](https://github.com/andycloke) - #329)
-   Fix YAxis using Unnecessary max/min value before definition ([@saji-ryu](https://github.com/saji-ryu) - #354)
-   prevent crash on undefined data ([@usrbowe](https://github.com/usrbowe) - #355)

*   remove eslint formatting in favor of prettier (easier to contribute)
*   remove circleci checks as they didn't add any value (easier to contribute)

# 5.2.0

-   Add `x(Min|Max)` and `y(Min|Max)`
-   Add `clamp(X|Y)` to use in conjunction with above props (default `false`)
-   Add xScale to StackedAreaChart along with (x|y)Accessor


storybook/stories/bar-stack/grouped.js
import React from 'react'
import { StackedBarChart, Grid } from 'rn-svg-charts'

class StackedBarChartExample extends React.PureComponent {
    render() {
        const data = [
            {
                data: [
                    {
                        month: new Date(2015, 0, 1),
                        apples: 3840,
                        bananas: 1920,
                        cherries: 960,
                        dates: 400,
                        oranges: 400,
                    },
                    {
                        month: new Date(2015, 1, 1),
                        apples: 1600,
                        bananas: 1440,
                        cherries: 960,
                        dates: 400,
                        oranges: 600,
                    },
                    {
                        month: new Date(2015, 2, 1),
                        apples: 640,
                        bananas: 960,
                        cherries: 3640,
                        dates: 400,
                        oranges: 800,
                    },
                    {
                        month: new Date(2015, 3, 1),
                        apples: 3320,
                        bananas: 480,
                        cherries: 640,
                        dates: 400,
                        oranges: 200,
                    },
                ],
            },
            {
                data: [
                    {
                        month: new Date(2015, 0, 1),
                        apples: 3840,
                        bananas: 1920,
                        cherries: 960,
                        dates: 400,
                        oranges: 400,
                    },
                    {
                        month: new Date(2015, 1, 1),
                        apples: 1600,
                        bananas: 1440,
                        cherries: 960,
                        dates: 400,
                        oranges: 600,
                    },
                    {
                        month: new Date(2015, 2, 1),
                        apples: 640,
                        bananas: 960,
                        cherries: 3640,
                        dates: 400,
                        oranges: 800,
                    },
                    {
                        month: new Date(2015, 3, 1),
                        apples: 3320,
                        bananas: 480,
                        cherries: 640,
                        dates: 400,
                        oranges: 200,
                    },
                ],
            },
            {
                data: [
                    {
                        month: new Date(2015, 0, 1),
                        apples: 3840,
                        bananas: 1920,
                        cherries: 960,
                        dates: 400,
                        oranges: 400,
                    },
                    {
                        month: new Date(2015, 1, 1),
                        apples: 1600,
                        bananas: 1440,
                        cherries: 960,
                        dates: 400,
                        oranges: 600,
                    },
                    {
                        month: new Date(2015, 2, 1),
                        apples: 640,
                        bananas: 960,
                        cherries: 3640,
                        dates: 400,
                        oranges: 800,
                    },
                    {
                        month: new Date(2015, 3, 1),
                        apples: 3320,
                        bananas: 480,
                        cherries: 640,
                        dates: 400,
                        oranges: 2500,
                    },
                ],
            },
        ]

        const colors = [['#8800cc', '#aa00ff'], ['#dd99ff', '#cc66ff'], ['#eeccff']]
        const keys = [['apples', 'bananas'], ['cherries', 'dates'], ['oranges']]

        return (
            <StackedBarChart
                style={{ height: 200 }}
                keys={keys}
                colors={colors}
                data={data}
                showGrid={false}
                contentInset={{ top: 30, bottom: 30 }}
            >
                <Grid />
            </StackedBarChart>
        )
    }
}

export default StackedBarChartExample
